### PR TITLE
external storage provision implement

### PR DESCRIPTION
**Pull request by soulseen**

### CODE REVIEW

File: controllers/storage/provision.go [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+			"accesskey":      []byte(m.HarborCluster.Spec.Storage.S3.Accesskey),
+			"secretkey":      []byte(m.HarborCluster.Spec.Storage.S3.Secretkey),
+			"regionendpoint": []byte(m.HarborCluster.Spec.Storage.S3.Regionendpoint),
+			"encrypt":        []byte(strconv.FormatBool(m.HarborCluster.Spec.Storage.S3.Encrypt)),
+			"keyid":          []byte(m.HarborCluster.Spec.Storage.S3.Keyid),
+			"secure":         []byte(strconv.FormatBool(m.HarborCluster.Spec.Storage.S3.Secure)),
+			"chunksize":      []byte(m.HarborCluster.Spec.Storage.S3.Chunksize),
+			"rootdirectory":  []byte(m.HarborCluster.Spec.Storage.S3.Rootdirectory),
+			"storageclass":   []byte(m.HarborCluster.Spec.Storage.S3.Storageclass),
+			"v4auth":         []byte(strconv.FormatBool(m.HarborCluster.Spec.Storage.S3.V4auth))},
```
Comment (wangcanfengxs) [inline controllers/storage/provision.go:66]: `Accesskey`，`Secretkey`, `Regionendpoint`... These words is not `CamelCase` naming.

File: controllers/storage/provision.go [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+)
+
+const (
+	Kind                        = "MinIOInstance"
+	ApiVersion                  = "miniooperator.min.io/v1beta1"
+	DefaultExternalSecretSuffix = "harbor-cluster-storage"
+	S3Secret                    = "s3Secret"
+	AzureSecret                 = "azureSecret"
+	GcsSecret                   = "gcsSecret"
+	SwiftSecret                 = "swiftSecret"
```
Comment (wangcanfengxs) [inline controllers/storage/provision.go:22]: No oos?

File: api/v1/harborcluster_types.go [hunk 3]
```diff
@@ -178,7 +187,7 @@ type Swift struct {
 	Authversion         int      `json:"authversion,omitempty"`
 	Endpointtype        string   `json:"endpointtype,omitempty"`
 	Tempurlcontainerkey bool     `json:"tempurlcontainerkey,omitempty"`
-	Tempurlmethods      []string `json:"tempurlmethods,omitempty"`
+	Tempurlmethods      string `json:"tempurlmethods,omitempty"`
```
Comment (wangcanfengxs) [inline api/v1/harborcluster_types.go:190]: Use `CamelCase` naming

File: controllers/storage/minio.go [hunk 4]
```diff
@@ -1,36 +1,151 @@
...
+		}
+		return minioReadyStatus(properties), nil
+	default:
+		// TODO
+		return nil, nil
+	}
 	return nil, nil
 }
 
-func (minio *MinIOReconciler) Provision(spec *goharborv1.HarborCluster) (*lcm.CRStatus, error) {
```
Comment (wangcanfengxs) [inline controllers/storage/minio.go:18]: You can not delete this method.

File: controllers/storage/provision.go [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+			"secretkey":           []byte(m.HarborCluster.Spec.Storage.Swift.Secretkey),
+			"authversion":         []byte(string(m.HarborCluster.Spec.Storage.Swift.Authversion)),
+			"endpointtype":        []byte(m.HarborCluster.Spec.Storage.Swift.Endpointtype),
+			"tempurlcontainerkey": []byte(strconv.FormatBool(m.HarborCluster.Spec.Storage.Swift.Tempurlcontainerkey)),
+			"tempurlmethods":      []byte(m.HarborCluster.Spec.Storage.Swift.Tempurlmethods),
+			"chunksize":           []byte(m.HarborCluster.Spec.Storage.Swift.Chunksize)},
+	}
+}
+
+func (m *MinIOReconciler) Provision() (*lcm.CRStatus, error) {
```
Comment (wangcanfengxs) [inline controllers/storage/provision.go:218]: How to use this method ?

Comment (soulseen) [reply controllers/storage/provision.go:219]: Provision inCluster are not implemented yet

File: controllers/storage/provision.go (hunk 1 above)
Comment (soulseen) [reply controllers/storage/provision.go:67]: ref [here](https://github.com/goharbor/harbor-operator/blob/c8e922bded7080598ee5d10861271b792a22086d/docs/installation/registry-storage-configuration.md#create-the-secret)

File: controllers/storage/provision.go (hunk 2 above)
Comment (soulseen) [reply controllers/storage/provision.go:22]: fixed

File: controllers/storage/minio.go (hunk 4 above)
Comment (soulseen) [reply controllers/storage/minio.go:18]: haven't deleted, just remove params of spec.

**Pull request merged**